File: src/graph/__tests__/fixtures/sample.rs
//! This is a sample Rust file for testing the code-graph extractor.

use std::collections::HashMap;

// 4. Declaration-order bug test
// impl before struct
impl Order {
    pub fn process(&self) {}
}

pub struct Order {
    pub id: u32,
}

/// A simple user struct
pub struct User {
    pub name: String,
    pub age: u32,
}

pub enum Role {
    Admin,
    Member,
}

pub trait Greeter {
    // 1. Trait methods test (no body)
    fn greet(&self) -> String;
}

impl Greeter for User {
    fn greet(&self) -> String {
        format!("Hello, my name is {}", self.name)
    }
}

pub const MAX_AGE: u32 = 120;
static GLOBAL_FLAG: bool = true;

pub mod admin {
    use super::*;
    
    pub fn is_admin(role: &Role) -> bool {
        match role {
            Role::Admin => true,
            _ => false,
        }
    }
}

// 3. Generic type parameters tests
pub fn make<T>() -> T {
    unimplemented!()
}

pub struct Box<T> {
    pub val: T,
}

pub trait Repo<T> {
    fn get(&self) -> T;
}

pub enum Result<T, E> {
    Ok(T),
    Err(E),
}

pub fn consume<T>(val: T) {}

pub fn create_user(name: String) -> User {
    let mut map = HashMap::new();
    map.insert(name.clone(), 1);
    
    admin::is_admin(&Role::Member);

    // 5. Duplicate call edges test
    // should only emit exactly ONE call to `make` and ONE call to `consume`
    consume(make::<i32>());

    // 2. Struct instantiation test
    // should emit exactly one `instantiates` to `User`
    User {
        name,
        age: 30,
    }
}

File: src/graph/__tests__/fixtures/sample.go
// Package main is a sample Go file for testing the code-graph extractor.

package main

import (
	"fmt"
	"strings"
)

// User represents a user in the system.
type User struct {
	Name string
	Age  int
}

type Role string

const (
	RoleAdmin Role = "admin"
	RoleMember Role = "member"
)

var globalFlag = true

// Greeter is an interface for greeting.
type Greeter interface {
	Greet() string
}

func (u *User) Greet() string {
	return fmt.Sprintf("Hello, my name is %s", u.Name)
}

func processOrder(order *Order) {
	fmt.Println("Processing order:", order.ID)
}

type Order struct {
	ID   int
	Items []string
}

func (o *Order) AddItem(item string) {
	o.Items = append(o.Items, item)
}

type Box[T any] struct {
	Value T
}

func makeBox[T any](val T) Box[T] {
	return Box[T]{Value: val}
}

func consume(val any) {}

func CreateUser(name string) *User {
	u := &User{Name: name, Age: 30}
	processOrder(&Order{ID: 1})
	u.Greet()
	consume(makeBox(42))
	return u
}

type Repo[T any] interface {
	Get() T
}
